staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go

```diff
@@ -237,14 +250,120 @@ func (d *validatingDispatcher) Dispatch(ctx context.Context, attr admission.Attr
 		return nil
 	}
 	if len(errs) > 1 {
-		for i := 1; i < len(errs); i++ {
-			// TODO: merge status errors; until then, just return the first one.
-			utilruntime.HandleError(errs[i])
-		}
+		return aggregateWebhookErrors(errs)
 	}
 	return errs[0]
 }
 
+func aggregateWebhookErrors(errs []error) error {
+	if len(errs) == 1 {
+		return errs[0]
+	}
+
+	errorsByReason := groupErrorsByReason(errs)
+
+	// When there are multiple groups of errors, we can only return one group.
+	// We use priorities to decide the group to return. The items are ordered from
+	// highest to lowest priority. We prioritize "actionable" errors, i.e. errors
+	// that can be resolved by some user action, like changing the resource.
+	for _, reason := range statusReasonsByPriorityDescending {
+		if group, exists := errorsByReason[reason]; exists {
+			mergedErr := mergeErrorsOfSameType(group)
+			if statusErr, ok := mergedErr.(*apierrors.StatusError); ok { //nolint:errorlint
+				return statusErr
+			}
+			return apierrors.NewInternalError(mergedErr)
+		}
+	}
+	return errs[0] // fallback
+}
+
+func groupErrorsByReason(errs []error) map[metav1.StatusReason][]error {
+	groups := make(map[metav1.StatusReason][]error)
+
+	for _, err := range errs {
+		// Log all original errors since they won't be directly surfaced to the client
+		utilruntime.HandleError(err)
+
+		if statusErr, ok := err.(*apierrors.StatusError); ok { //nolint:errorlint
+			reason := statusErr.Status().Reason
+			groups[reason] = append(groups[reason], err)
+		} else {
+			// For non-StatusError, categorize as InternalError
+			groups[metav1.StatusReasonInternalError] = append(groups[metav1.StatusReasonInternalError], err)
+		}
+	}
+
+	return groups
+}
+
+func mergeErrorsOfSameType(errs []error) error {
+	if len(errs) == 1 {
+		return errs[0]
+	}
+
+	var firstStatusErr *apierrors.StatusError
+	var details []metav1.StatusDetails
+	var messages []string
+
+	for _, err := range errs {
+		if statusErr, ok := err.(*apierrors.StatusError); ok { //nolint:errorlint
+			if firstStatusErr == nil {
+				firstStatusErr = statusErr
+			}
+			if statusErr.Status().Details != nil {
+				details = append(details, *statusErr.Status().Details)
+			}
+			messages = append(messages, statusErr.Status().Message)
+		}
+	}
+
+	// If no StatusError found, return the first error
+	if firstStatusErr == nil {
+		return errs[0]
+	}
+
+	return &apierrors.StatusError{
+		ErrStatus: metav1.Status{
+			Code:    firstStatusErr.Status().Code,
+			Reason:  firstStatusErr.Status().Reason,
```

Comment:
We don't do much validation of the error statuses returned by validation webhooks today, so I'd expect the detailed values for status fields can probably be all over the map
At first glance, I think we should only merge errors whose code, reason, details.{name,group,kind,uid} fields are identical
I'm not sure it's reasonable to assume the code, reason, details.{name,group,kind,uid} fields from the first encountered error are sensible to merge messages / causes from subsequent errors into if those don't match

Reply:
You're absolutely right! I've updated the implementation to only merge errors when all the key fields (code, reason, details.{name,group,kind,uid}) are identical.